bugsnag-android-core/src/main/java/com/bugsnag/android/internal/DocumentPath.kt

```diff
@@ -0,0 +1,209 @@
+package com.bugsnag.android
+
+/**
+ * DocumentPath records a path into a hierarchical acyclic document, and can be used to modify
+ * documents by navigating down the DAG and then applying a modification at that level.
+ *
+ * Paths are explained as follows in PLAT-6353:
+ *
+ * Paths follow a dotted notation similar to internet hostnames, such that each path component
+ * is separated by a dot '.' If a literal dot is required, the path supports an escaping
+ * mechanism using the backslash character:
+ *
+ * \. = literal dot
+ * \\ = literal backslash
+ *
+ * Path components are either integers or non-integers. Integers refer to the previous path
+ * component as a list, providing the index to look up. Non-integers refer to the previous path
+ * component as a map, providing the key to look up (for example, in the path "a.b.3", "a" is
+ * treated as a map and "b" is treated as a list, so the path refers to document root with map
+ * key "a", which is then indexed with map key "b", which is then indexed with list index 3).
+ *
+ * Any nonexistent parts of the document are created on-demand as they are encountered while
+ * resolving paths (either as a list in the case of integer path components, or as a map for
+ * non-integers). In the case of lists, only index 0 of a nonexistent list is valid (in which
+ * case the list will be created, and the value stored at index 0).
+ *
+ * As well, the following special rules apply:
+ *
+ * - An empty path component refers to the document root, and causes the value to replace
+ * the entire document.
+ * - The integer path component -1 refers to the last index in the array, or index 0 if the
+ * array is empty or nonexistent.
+ * - If a path ends in a dot (e.g. events.exceptions.stacktrace.), It means that the last
+ * path entry (e.g. stacktrace) is to be treated as a list, and the value is to be appended
+ * to that list.
+ *
+ * The document to be modified must be serializable into JSON. The following types are supported:
+ * - null
+ * - boolean
+ * - integer (max 15 digits)
+ * - float
+ * - string
+ * - List(Object)
+ * - Map(String, Object)
+ */
+class DocumentPath(path: String) {
```

Comment:
If I mark it private, the benchmark tests can't find it.

Reply:
Can the benchmark test also be in the `internal` package?

Reply:
The benchmark test is in a different module so `internal` classes won't be accessible. It seems preferable to make this class `internal` to avoid exposing it to users and delete the benchmark test.

Reply:
I see. Though in that case, not being able to easily benchmark code that we expect to be on the hottest path in the library seems like a greater risk than the class being arbitrarily accessible.

Reply:
The existing breadcrumb/metadata benchmarks should give some coverage once they've been updated to add journal changes. Perhaps it would be more valuable to keep `Journal` public in #1307 and write the benchmark tests against that?

Reply:
Is it really so bad to have this code accessible? It's in "internal", so anyone who calls such internal code will know that they're courting ruin.

I'm all for solid APIs that don't contain confusion and footguns from unnecessarily exposed interfaces, but "idiot proofing" hits diminishing returns after awhile, and eventually restricts our own usage in testing...